Bump python app

sicheste opened the pull request:
Using setup-python v5 and Python 3.13.

<!--
IMPORTANT:

This repository contains configuration for what users see when they click on the `Actions` tab and the setup page for Code Scanning.

It is not:
* A playground to try out scripts
* A place for you to create a workflow for your repository
-->

## Pre-requisites

- [ ] Prior to submitting a new workflow, please apply to join the GitHub Technology Partner Program: [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner).

---

### **Please note that at this time we are only accepting new starter workflows for Code Scanning. Updates to existing starter workflows are fine.**

---

## Tasks

**For _all_ workflows, the workflow:**

- [x] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [x] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [x] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [x] Should include comments in the workflow for any parts that are not obvious or could use clarification.
- [x] Should specify least privileged [permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) for `GITHUB_TOKEN` so that the workflow runs successfully.

**For _CI_ workflows, the workflow:**

- [x] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [x] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [x] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [ ] Packaging workflows should run on `release` with `types: [ created ]`.
- [ ] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**For _Code Scanning_ workflows, the workflow:**

- [ ] Should be preserved under [the `code-scanning` directory](https://github.com/actions/starter-workflows/tree/main/code-scanning).
- [ ] Should include a matching `code-scanning/properties/*.properties.json` file (for example, [`code-scanning/properties/codeql.properties.json`](https://github.com/actions/starter-workflows/blob/main/code-scanning/properties/codeql.properties.json)), with properties set as follows:
  - [ ] `name`: Name of the Code Scanning integration.
  - [ ] `creator`: Name of the organization/user producing the Code Scanning integration.
  - [ ] `description`: Short description of the Code Scanning integration.
  - [ ] `categories`: Array of languages supported by the Code Scanning integration.
  - [ ] `iconName`: Name of the SVG logo representing the Code Scanning integration. This SVG logo must be present in [the `icons` directory](https://github.com/actions/starter-workflows/tree/main/icons).
- [ ] Should run on `push` to `branches: [ $default-branch, $protected-branches ]` and `pull_request` to `branches: [ $default-branch ]`. We also recommend a `schedule` trigger of `cron: $cron-weekly` (for example, [`codeql.yml`](https://github.com/actions/starter-workflows/blob/c59b62dee0eae1f9f368b7011cf05c2fc42cf084/code-scanning/codeql.yml#L14-L21)).

**Some general notes:**

- [x] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [x] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [x] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [x] Automation and CI workflows cannot be dependent on a paid service or product.

sicheste committed:
use current setup-python v5

sicheste committed:
use Python 3.13

reactive-firewall commented on ci/python-app.yml:
```diff
@@ -19,10 +19,10 @@ jobs:
+      uses: actions/setup-python@v5
       with:
-        python-version: "3.10"
+        python-version: "3.13"
```
LGTM! with Conditions:

 * verified improvement of `actions/setup-python@v5` (from `v3`).
 * Meta-data is consistent and matches the Python version used.
   <details><summary>NIT</summary>

     > [!TIP]
     > knit-picky or out-of-scope:
     > Consider improving with a single variable to set the python version and
     > access it for both name and python-version via GHA contexts.
     > See [Documentation](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs) for details.)

   </details>
   <details><summary>NIT - Consider Python `3.12` might be more appropriate than `3.13` for GHA</summary>

     > [!TIP]
     > knit-picky or out-of-scope:
     > Looks like version 3.12 will be the cached version (while `3.13` **is** the
     > latest it will probably be slower for some GitHub Actions)
     > see [https://github.com/actions/runner-images/issues/10636](https://github.com/actions/runner-images/issues/10636)
     > and note the cached versions for python.
     >
     > Should still work though; Just slightly slower.

   </details>

**Hope this helps** :bow:

sicheste replied:
Regarding the variables:
I've checked some other files. Have not seen that it was done like this somehwere. For example the android, ant and maven templates are using the version number in the step name and for the version requirement. But always explicitly written. So yes, technically a considerable idea, but have not seen that it is done somewhere else like this.

Using Python 3.12 instead of 3.13 makes sense and I have changed.

cclauss commented on ci/python-app.yml:
```diff
@@ -19,10 +19,10 @@ jobs:
+      uses: actions/setup-python@v5
       with:
-        python-version: "3.10"
+        python-version: "3.13"
```
```suggestion
        python-version: "3.x"
```
https://devguide.python.org/versions/#supported-versions
Testing on the latest version of Python may catch new language incompatibilities before most users notice them.

reactive-firewall replied:
> [!WARNING]
> 🙅🏻 Syntax error

@cclauss [that suggestion](https://github.com/actions/starter-workflows/pull/2639#discussion_r1874591190) would regress the action to a broken state. Please re-read the docs you correctly mentioned.

I'll also re-iterate what I previously marked knitpicky as it now directly applies to this, see below:

Consider Python `3.12` might be more appropriate than `3.13` for GHA
> [!TIP]
> knit-picky or out-of-scope:
Looks like version 3.12 will be the cached version (while 3.13 is the
latest it will probably be slower for some GitHub Actions)
see https://github.com/actions/runner-images/issues/10636
and note the cached versions for python.

> Should still work though; Just slightly slower.

In conclusion; pick one (exactly one):
 * `3.12` - for tuned value based on GitHub Action's runners.
 * `3.13` - the latest available version of Python as per your link; "Should still work though; Just slightly slower" due to not being cached.
 * any other less current value in the form defined by the [`actions/setup-pythonv@5`](https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md) (`3.9.20`|`3.10`|`3.11`) (included for completeness)

> [!IMPORTANT]
> This is not to say your rational is wrong, (only the implementation)

To also implement checking for the latest version please refer to [this guide](https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#check-latest-version)

or as code:
```python
# Rest of config is unchanged
  - uses: actions/setup-python@v5
    with:
      python-version: '3.13'
      check-latest: true
```

sicheste replied:
I like the idea using a cached Python version by default. So I switched to 3.12.

sicheste replied:
In python-publish it is given with `3.x`. So would not be the first template when setting it to .x here.

reactive-firewall replied:
> In python-publish it is given with `3.x`. So would not be the first template when setting it to .x here.


## Oops 🙊

My oversight; Correction: it is a runtime error, not a syntax error. I'll remove the incorrect and confusing phrasing from my previous comment.

For those curious, my full rational,
<details><summary>TL;DR</summary>

This gave me pause, at first, as you are quite correct, many (e.g. [`python-publish`](https://github.com/actions/starter-workflows/blob/main/ci/python-publish.yml#L27), and even the GitHub guides use `3.x` all over, and only [sometimes do they document why](https://docs.github.com/en/actions/use-cases-and-examples/building-and-testing/building-and-testing-python#using-a-specific-python-version)). After some digging all I could find was:
```yaml
name: Python package

on: [push]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
      - name: Set up Python
        # This is the version of the action for setting up Python, not the Python version.
        uses: actions/setup-python@v5
        with:
          # Semantic version range syntax or exact version of a Python version
          python-version: '3.x'
          # Optional - x64 or x86 architecture, defaults to x64
          architecture: 'x64'
      # You can test your matrix by printing the current Python version
      - name: Display Python version
        run: python -c "import sys; print(sys.version)"
```

which makes sense if you are familiar with [Semantic Version range syntax](https://semver.org/#semantic-versioning-specification-semver) which boils down to [almost the same thing](https://regex101.com/r/vkijKf/1/) as I expected (e.g. PEP-440 and the current [Python packaging version spec](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-specifiers))
The notable difference is `3.x` is invalid as a **Semantic Version** but is considered a local version `x` of version with Major `3` resulting in 

so I checked [the code](https://github.com/actions/setup-python/blob/main/action.yml#L7C5-L7C150) which documents the same thing, (albeit without the example placeholder of `3.x`). So I wanted to see what it did with the version, and deep inside the massive [`dist/index.js`](https://raw.githubusercontent.com/actions/setup-python/refs/heads/main/dist/setup/index.js) of the `python-setup` action is a function `resolveVersionInput()` to resolve the input range that would essentially return `['3.x']`, from there it is used by the `run()` function to attempt to download it via the function `useCpythonVersion` (because `3.x` will not match other cases like 'PyPy' etc.) there it finally is parsed (and unless `python3.x` cpython is already installed, will fail and fall-back on a default of `void 0` (e.g. `None` version), if there is a `python3.x` installed on the GHA image is most likely the documented default of `3.12`) resulting eventually in a harmless runtime-error, and reporting the actually installed version via the output `python-version` (same name as input) which would currently be `3.12`.

In conclusion: `3.x` is still not advised, as omitting the value would be preferred to achieve the same behavior;
the original `3.13` should work;
and I still recommend the cached `3.12`.

### Informative references
 * `actions/setup-python@v5`/[`dist/index.js`](https://raw.githubusercontent.com/actions/setup-python/refs/heads/main/dist/setup/index.js) - the code.
 * [using-a-specific-python-version](https://docs.github.com/en/actions/use-cases-and-examples/building-and-testing/building-and-testing-python#using-a-specific-python-version) - GitHub guide on python-setup action topic
 * [python-action](https://github.com/actions/setup-python) and [docs](https://github.com/actions/setup-python#basic-usage)
 * [semantic versioning specification](https://semver.org/#semantic-versioning-specification-semver)
 * [semantic versioning reg-Ex](https://regex101.com/r/vkijKf/1/) - spec recommended way to test version format
 * [Python versioning specification](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-specifiers)
</details>

---

## Summary

In summary: This topic has been thoroughly explored now, I agree with @cclauss that just using the latest stable would be a fine alternative, but not better than the other already mentioned options.

@sicheste `3.12` for leveraging caching in GHAs, looks good, 👍🏻

sicheste replied:
Thanks for the long analysis and explanation.

So, now someone can merge it :)

sicheste committed:
Use Python 3.12.

Python 3.12 is cached by the runners. 3.13 has to be downloaded by the runners every run again, so using 3.12 by default speeds up the execution.

reactive-firewall commented on ci/python-app.yml:
**LGTM** 👍🏻

* consistent meta-data and values
* versions are stable and current

JamesonDiamond1 replied:
@JamesonDiamond1

JamesonDiamond1 reviewed (approved):
@JamesonDiamond1

reactive-firewall reviewed (approved):
LGTM :+1:
 * versions are stable and current

JamesonDiamond1 reviewed (approved):
its not uplaoding a complete file

sicheste commented:
outdated
